facebookads/api.py

```diff
@@ -500,25 +500,26 @@ def execute(self):
         retry_indices = []
 
         for index, response in enumerate(responses):
-            if response:
-                body = response.get('body')
-                code = response.get('code')
-                headers = response.get('headers')
-
-                inner_fb_response = FacebookResponse(
-                    body=body,
-                    headers=headers,
-                    http_status=code,
-                    call=self._batch[index],
-                )
-
-                if inner_fb_response.is_success():
-                    if self._success_callbacks[index]:
-                        self._success_callbacks[index](inner_fb_response)
-                elif self._failure_callbacks[index]:
-                    self._failure_callbacks[index](inner_fb_response)
-            else:
+            if not response:
                 retry_indices.append(index)
+                continue
+
+            inner_fb_response = FacebookResponse(
+                body=response.get('body'),
+                headers=response.get('headers'),
+                http_status=response.get('code'),
+                call=self._batch[index],
+            )
+            if inner_fb_response.is_success() and self._success_callbacks[index]:
+                self._success_callbacks[index](inner_fb_response)
+            else:
+                # retry transient errors
+                if inner_fb_response.is_transient():
+                    retry_indices.append(index)
+
+                # execute failure callbacks on error, even if transient
+                if self._failure_callbacks[index]:
```

Comment:
I see and I agree that option 3 is better. But I sense that there would be a lot of changes to do: every time a batch request is done in the SDK, we need to add the transient_error_callback argument. The signatures of all functions would change, etc... and since the unit-test coverage is not that good, I don't think it's great to take that risk (but you're the maintainer so you choose ^^)

I think that we could be smarter and do that at the batch object level. For instance:
```python
class FacebookAdsApiBatch(object):
    def __init__(self, api, transient_policy=None):
        self._api = api
        ...
        # it's part of a enum like ["SUCCESS", "SILENT", "FAILURE"]
        self.transient_policy = transient_policy
```

And in the `execute()` method we would call the right callback. By default, it's silent, so the current behaviour doesn't change. The good thing about this is that it doesn't change all the function signature, there's just a new optional parameter to the batch initialization. What do you think?